src/noteburst/jupyterclient/jupyterlab.py

```diff
@@ -0,0 +1,708 @@
+            "channel": "shell",
+            "content": {
+                "code": code,
+                "silent": False,
+                "store_history": False,
+                "user_expressions": {},
+                "allow_stdin": False,
+            },
+            "metadata": {},
+            "buffers": {},
+        }
+
+        await self.websocket.send(json.dumps(msg))
+
+        result = ""
+        async for message in self.websocket:
+            try:
+                result += await self._process_run_python_message(
+                    message, msg_id, code
+                )
+            except CodeExecutionError as e:
+                if result:
+                    error_message = f"{result}\nError:\n{str(e.error)}"
+                else:
+                    error_message = str(e)
+                raise CodeExecutionError(
+                    username=e.username,
+                    code=e.code,
+                    code_type=e.code_type,
+                    error=error_message,
+                    status=e.status,
+                )
+            except JupyterLabSessionStreamComplete:
+                break
+
+        return result
+
+    async def _process_run_python_message(
+        self, raw_message: websockets.typing.Data, msg_id: str, code: str
+    ) -> str:
+        """Process an individual message received from the websocket,
+        initiated from `run_python`.
+
+        This method returns incremental fragments of the result string. If
+        the message is an ``execute_reply`` with status of ``ok``, then this
+        method raises `JupyterLabSessionStreamComplete`, which indicates that
+        the result is complete and the message iteration can be stopped.
+        """
+        if isinstance(raw_message, bytes):
+            raw_text = raw_message.decode("utf-8")
+        else:
+            raw_text = raw_message
+        response_data = json.loads(raw_text)
+
+        self.logger.debug(f"Received kernel message: {response_data}")
+        msg_type = response_data["msg_type"]
+
+        if response_data.get("parent_header", {}).get("msg_id") != msg_id:
+            # Ignore messages not intended for us. The web socket is
+            # rather chatty with broadcast status messages.
+            return ""
+        elif msg_type == "error":
+            error = "".join(response_data["content"]["traceback"])
+            raise CodeExecutionError(
+                username=self.username, code=code, error=error
+            )
+        elif msg_type == "stream":
+            return response_data["content"]["text"]
+        elif msg_type == "execute_reply":
+            status = response_data["content"]["status"]
+            if status == "ok":
+                raise JupyterLabSessionStreamComplete
+            else:
+                raise CodeExecutionError(
+                    username=self.username, code=code, status=status
+                )
+        return ""  # shouldn't generally be reached
```

Comment:
This was a problem in the original mobu code, but it might be worth logging some sort of message here about an unexpected `msg_type`, since otherwise it's going to be hard to understand what's going on when some future version of the lab starts using a new message type for something we care about.